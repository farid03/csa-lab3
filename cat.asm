section data:
buffer: 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0
aboba: 1
str: 'aaaa', 0

section text:
    _start:
        add r2,   [r0]  , buffer
        add r3,r0,6969 ; STDIN
    read:
        lw r1,r3
        sw r2,r1
        beq r1,r0,finish_read
        add r2,r2,1
        jmp read
    finish_read:
        add r2,r0,buffer
        add r3,r0,9696 ; STDOUT

    write:
        lw r1,r2
        sw r3,r1
        beq r1,r0,end ; f
        add r2,r2,1
        jmp write
    end:
        halt
