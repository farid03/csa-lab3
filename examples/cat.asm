section data:
buffer: 0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0

section text:
    _start:
        add r2,r0,buffer
    read:
        in r1, 0
        sw r2,r1
        beq r1,r0,finish_read ; пока не достигнем нуль-терминатора
        add r2,r2,1
        jmp read
    finish_read:
        add r2,r0,buffer

    write:
        lw r1,r2
        out r1, 1
        beq r1,r0,end
        add r2,r2,1
        jmp write
    end:
        halt