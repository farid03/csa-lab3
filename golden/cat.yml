source: |-
  ,[.,]
input: |-
  foo
code: |-
  [
      {
          "opcode": "input",
          "term": [
              1,
              1,
              ","
          ]
      },
      {
          "opcode": "jz",
          "arg": 5,
          "term": [
              1,
              2,
              "["
          ]
      },
      {
          "opcode": "print",
          "term": [
              1,
              3,
              "."
          ]
      },
      {
          "opcode": "input",
          "term": [
              1,
              4,
              ","
          ]
      },
      {
          "opcode": "jmp",
          "arg": 1,
          "term": [
              1,
              5,
              "]"
          ]
      },
      {
          "opcode": "halt"
      }
  ]
output: |
  source LoC: 1 code instr: 6
  ============================================================
  foo
  instr_counter:  11 ticks: 21
log: |
  DEBUG    root:machine.py:257 {TICK: 0, PC: 0, ADDR: 0, OUT: 0, ACC: 0} input  (',' @ 1:1)
  DEBUG    root:machine.py:117 input: 'f'
  DEBUG    root:machine.py:263 {TICK: 2, PC: 1, ADDR: 0, OUT: 102, ACC: 0} jz 5 ('[' @ 1:2)
  DEBUG    root:machine.py:263 {TICK: 4, PC: 2, ADDR: 0, OUT: 102, ACC: 102} print  ('.' @ 1:3)
  DEBUG    root:machine.py:92 output: '' << 'f'
  DEBUG    root:machine.py:263 {TICK: 6, PC: 3, ADDR: 0, OUT: 102, ACC: 102} input  (',' @ 1:4)
  DEBUG    root:machine.py:117 input: 'o'
  DEBUG    root:machine.py:263 {TICK: 8, PC: 4, ADDR: 0, OUT: 111, ACC: 102} jmp 1 (']' @ 1:5)
  DEBUG    root:machine.py:263 {TICK: 9, PC: 1, ADDR: 0, OUT: 111, ACC: 102} jz 5 ('[' @ 1:2)
  DEBUG    root:machine.py:263 {TICK: 11, PC: 2, ADDR: 0, OUT: 111, ACC: 111} print  ('.' @ 1:3)
  DEBUG    root:machine.py:92 output: 'f' << 'o'
  DEBUG    root:machine.py:263 {TICK: 13, PC: 3, ADDR: 0, OUT: 111, ACC: 111} input  (',' @ 1:4)
  DEBUG    root:machine.py:117 input: 'o'
  DEBUG    root:machine.py:263 {TICK: 15, PC: 4, ADDR: 0, OUT: 111, ACC: 111} jmp 1 (']' @ 1:5)
  DEBUG    root:machine.py:263 {TICK: 16, PC: 1, ADDR: 0, OUT: 111, ACC: 111} jz 5 ('[' @ 1:2)
  DEBUG    root:machine.py:263 {TICK: 18, PC: 2, ADDR: 0, OUT: 111, ACC: 111} print  ('.' @ 1:3)
  DEBUG    root:machine.py:92 output: 'fo' << 'o'
  DEBUG    root:machine.py:263 {TICK: 20, PC: 3, ADDR: 0, OUT: 111, ACC: 111} input  (',' @ 1:4)
  WARNING  root:machine.py:265 Input buffer is empty!
  INFO     root:machine.py:268 output_buffer: 'foo'
